src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -56,8 +56,320 @@ private static unsafe void ThrowExceptionEncryptDecryptFail(string fullPath)
+                                0,
+                                out _,
+                                IntPtr.Zero))
+                        {
+                            throw new Exception("R");
+                            //return false;
+                        }
+                    }
+
+                    // Read the file attributes.
+                    Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA sourceFileInformation = default;
+                    int error = FillAttributeInfo(sourceHandle, ref sourceFileInformation);
+                    Debug.Assert(error == 0); //todo: is there a good reason for this to fail? if so we should handle it properly, not by a Debug.Assert
+
+                    // Get file size.
+                    long sourceSize = (long)(((ulong)sourceFileInformation.nFileSizeHigh << 32) | sourceFileInformation.nFileSizeLow);
+
+                    // Set length of destination to same as source.
+                    Interop.Kernel32.FILE_END_OF_FILE_INFO eofInfo;
+                    eofInfo.EndOfFile = sourceSize;
+                    if (!Interop.Kernel32.SetFileInformationByHandle(destinationHandle, Interop.Kernel32.FileEndOfFileInfo, &eofInfo, (uint)sizeof(Interop.Kernel32.FILE_END_OF_FILE_INFO)))
+                    {
+                        throw new Exception("S");
+                        //return false;
+                    }
+
+                    // Copy all of the blocks.
+                    Interop.Kernel32.DUPLICATE_EXTENTS_DATA duplicateExtentsData;
+                    duplicateExtentsData.FileHandle = sourceHandle.DangerousGetHandle();
+                    // ReFS requires that cloned regions reside on a disk cluster boundary.
+                    long clusterSizeMask = clusterSize - 1;
+                    long fileSizeRoundedUpToClusterBoundary = (sourceSize + clusterSizeMask) & ~clusterSizeMask;
+                    long sourceOffset = 0;
+                    while (sourceOffset < sourceSize)
+                    {
+                        duplicateExtentsData.SourceFileOffset = sourceOffset;
+                        duplicateExtentsData.TargetFileOffset = sourceOffset;
+                        const long MaxChunkSize = 1L << 31; // Each cloned region must be < 4GiB in length. Use a smaller default (2GiB).
+                        long thisChunkSize = Math.Min(fileSizeRoundedUpToClusterBoundary - sourceOffset, MaxChunkSize);
+                        duplicateExtentsData.ByteCount = thisChunkSize;
+
+                        if (!Interop.Kernel32.DeviceIoControl(
+                            destinationHandle,
+                            Interop.Kernel32.FSCTL_DUPLICATE_EXTENTS_TO_FILE,
+                            &duplicateExtentsData,
+                            (uint)sizeof(Interop.Kernel32.DUPLICATE_EXTENTS_DATA),
+                            null,
+                            0,
+                            out _,
+                            IntPtr.Zero))
```

Comment:
Reminder for future me: could we use asynchronous IO here to copy all the sections simultaneously?